storm-client/src/jvm/org/apache/storm/security/auth/workertoken/WorkerTokenAuthorizer.java

```diff
@@ -0,0 +1,139 @@
+                throw new RuntimeException(e);
+            }
+        }
+        return state;
+    }
+
+    private final LoadingCache<WorkerTokenInfo, PrivateWorkerKey> keyCache;
+
+    /**
+     * Constructor.
+     * @param conf the daemon config for the server.
+     * @param connectionType the type of connection we are authorizing.
+     */
+    public WorkerTokenAuthorizer(Map<String, Object> conf, ThriftConnectionType connectionType) {
+        this(connectionType.getWtType(), buildStateIfNeeded(conf, connectionType));
+    }
+
+    @VisibleForTesting
+    WorkerTokenAuthorizer(final WorkerTokenServiceType serviceType, final IStormClusterState state) {
+        LoadingCache<WorkerTokenInfo, PrivateWorkerKey> tmpKeyCache = null;
+        if (state != null) {
+            tmpKeyCache =
+                CacheBuilder.newBuilder()
+                    .maximumSize(2_000)
+                    .expireAfterWrite(2, TimeUnit.HOURS)
+                    .build(new CacheLoader<WorkerTokenInfo, PrivateWorkerKey>() {
+
+                        @Override
+                        public PrivateWorkerKey load(WorkerTokenInfo wtInfo) {
+                            return state.getPrivateWorkerKey(serviceType,
+                                wtInfo.get_topologyId(),
+                                wtInfo.get_secretVersion());
+                        }
+                    });
+        }
+        keyCache = tmpKeyCache;
+    }
+
+    @VisibleForTesting
+    byte[] getSignedPasswordFor(byte[] user, WorkerTokenInfo deser) {
+        assert keyCache != null;
+
+        if (deser.is_set_expirationTimeMillis() && deser.get_expirationTimeMillis() <= Time.currentTimeMillis()) {
+            throw new IllegalArgumentException("Token is not valid, token has expired.");
+        }
+
+        PrivateWorkerKey key = keyCache.getUnchecked(deser);
+        if (key == null) {
+            throw new IllegalArgumentException("Token is not valid, private key not found.");
+        }
+
+        if (key.is_set_expirationTimeMillis() && key.get_expirationTimeMillis() <= Time.currentTimeMillis()) {
+            throw new IllegalArgumentException("Token is not valid, key has expired.");
+        }
+
+        return WorkerTokenSigner.createPassword(user, new SecretKeySpec(key.get_key(), WorkerTokenSigner.DEFAULT_HMAC_ALGORITHM));
+    }
+
+    @Override
+    public Optional<char[]> getPasswordFor(String userName) {
+        if (keyCache == null) {
+            return Optional.empty();
+        }
+        try {
+            byte[] user = Base64.getDecoder().decode(userName);
+            WorkerTokenInfo deser = Utils.deserialize(user, WorkerTokenInfo.class);
+            byte[] password = getSignedPasswordFor(user, deser);
+            return Optional.of(Base64.getEncoder().encodeToString(password).toCharArray());
+        } catch (Exception e) {
+            LOG.debug("Could not decode {}, might just be a plain digest request...", userName, e);
```

Comment:
The token exceptions above aren't worth an info line if they occur?

Reply:
The issue is because in some cases we have to support both a regular DIGEST-MD5 and a token DIGEST-MD5.  In the regular case the user name and password have no special translation.  They are just read from a file.  In the token case they are translated.

The way we distinguish between the two is to try and deserialize it here, and if it works then it is a token, else assume it is not a token.